src/ZODB/Connection.py

```diff
@@ -101,11 +103,10 @@ def __init__(self, db, cache_size=400, before=None, cache_size_bytes=0):
         # Multi-database support
         self.connections = {self._db.database_name: self}
 
-        storage = db._mvcc_storage
         if before:
-            storage = storage.before_instance(before)
+            storage = HistoricalStorageAdapter(db.storage, before)
```

Comment:
Seems reasonable to me. 

The try/catch has to stay for BWC for now, but maybe one day IMVCCStorage should include `before_instance`, which RelStorage would implement pretty much exactly like the function you defined.